sdks/java/core/src/test/java/org/apache/beam/sdk/io/TFRecordIOTest.java

```diff
@@ -0,0 +1,315 @@
+    assertThat(displayData, hasDisplayItem("validation", false));
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testReadOne() throws Exception {
+    runTestRead(FOO_RECORD_BASE64, FOO_RECORDS);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testReadTwo() throws Exception {
+    runTestRead(FOO_BAR_RECORD_BASE64, FOO_BAR_RECORDS);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWriteOne() throws Exception {
+    runTestWrite(FOO_RECORDS, FOO_RECORD_BASE64);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWriteTwo() throws Exception {
+    runTestWrite(FOO_BAR_RECORDS, FOO_BAR_RECORD_BASE64, BAR_FOO_RECORD_BASE64);
+  }
+
+  private void runTestRead(String base64, String[] expected) throws IOException {
+    File tmpFile = Files.createTempFile(tempFolder, "file", ".tfrecords").toFile();
+    String filename = tmpFile.getPath();
+
+    System.out.println(tmpFile);
+    FileOutputStream fos = new FileOutputStream(tmpFile);
+    fos.write(BaseEncoding.base64().decode(base64));
+    fos.close();
+
+    TFRecordIO.Read.Bound read = TFRecordIO.Read.from(filename);
+    PCollection<String> output = p.apply(read).apply(ParDo.of(new ByteArrayToString()));
+
+    PAssert.that(output).containsInAnyOrder(expected);
+    p.run();
+  }
+
+  private void runTestWrite(String[] elems, String ...base64) throws IOException {
+    File tmpFile = Files.createTempFile(tempFolder, "file", ".tfrecords").toFile();
+    String filename = tmpFile.getPath();
+
+    PCollection<byte[]> input = p.apply(Create.of(Arrays.asList(elems)))
+        .apply(ParDo.of(new StringToByteArray()));
+
+    TFRecordIO.Write.Bound write = TFRecordIO.Write.to(filename).withoutSharding();
+    input.apply(write);
+
+    p.run();
+
+    FileInputStream fis = new FileInputStream(tmpFile);
+    String written = BaseEncoding.base64().encode(ByteStreams.toByteArray(fis));
+    // bytes written may vary depending the order of elems
+    assertThat(written, isIn(base64));
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void runTestRoundTrip() throws IOException {
+    runTestRoundTrip(LARGE, 10, ".tfrecords", NONE, NONE);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void runTestRoundTripWithEmptyData() throws IOException {
+    runTestRoundTrip(EMPTY, 10, ".tfrecords", NONE, NONE);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void runTestRoundTripWithOneShards() throws IOException {
+    runTestRoundTrip(LARGE, 1, ".tfrecords", NONE, NONE);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void runTestRoundTripWithSuffix() throws IOException {
+    runTestRoundTrip(LARGE, 10, ".suffix", NONE, NONE);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void runTestRoundTripGzip() throws IOException {
+    runTestRoundTrip(LARGE, 10, ".tfrecords", GZIP, GZIP);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void runTestRoundTripZlib() throws IOException {
```

Comment:
Add following tests related to validation from https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/tfrecordio_test.py (and corresponding validation logic to reader) ?

test_read_record_invalid_record()
test_read_record_invalid_length_mask()
test_read_record_invalid_data_mask()